[JSC] Fractional part of an ISO8601 Duration string must not be interpreted as floating-point

rkirsling opened the pull request:
#### 92e204d23f40144bdb3908c60f2d4fb66e5f6ae9
<pre>
[JSC] Fractional part of an ISO8601 Duration string must not be interpreted as floating-point
<a href="https://bugs.webkit.org/show_bug.cgi?id=241940">https://bugs.webkit.org/show_bug.cgi?id=241940</a>

Reviewed by Yusuke Suzuki.

The last time part of an ISO8601 Duration string is allowed to have 1 to 9 digits beyond the decimal point.
If this fractional part is interpreted using floating-point arithmetic, off-by-one errors can occur.
(<a href="https://tc39.es/proposal-temporal/#sec-temporal-parsetemporaldurationstring)">https://tc39.es/proposal-temporal/#sec-temporal-parsetemporaldurationstring)</a>

Contrary to the spec, test262 tests had been expecting IEEE-754-aligned results.
Tests were corrected in tc39/test262#3544; this patch corrects our implementation.

* JSTests/test262/expectations.yaml:
Mark 16 test cases as passing.

* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::handleFraction):

Canonical link: <a href="https://commits.webkit.org/251809@main">https://commits.webkit.org/251809@main</a>
</pre>

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -70,50 +70,37 @@ static int32_t parseDecimalInt32(const CharType* characters, unsigned length)
+        padded[i] = fractionString[i];
 
-    double fraction = factor * parseInt(fractionString, 10) / std::pow(10, fractionString.length());
+    int64_t fraction = factor * parseDecimalInt32(padded.data(), 9);
```
Let's cast `factor` or parseDecimalInt32 result with `static_cast<int64_t>(...)` to ensure that multiplication is done in int64_t.

rkirsling replied:
Oh, good catch!

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -70,50 +70,37 @@ static int32_t parseDecimalInt32(const CharType* characters, unsigned length)
-        duration.setMinutes(std::trunc(fraction));
-        fraction = std::fmod(fraction, 1);
+        duration.setMinutes(fraction / nsPerSecond);
+        fraction %= nsPerSecond;
```
nsPerMinute?

rkirsling replied:
Yeah, this needs an alias somehow.

Constellation reviewed (approved):
r=me with comments

rkirsling committed:
[JSC] Fractional part of an ISO8601 Duration string must not be interpreted as floating-point
https://bugs.webkit.org/show_bug.cgi?id=241940

Reviewed by Yusuke Suzuki.

The last time part of an ISO8601 Duration string is allowed to have 1 to 9 digits beyond the decimal point.
If this fractional part is interpreted using floating-point arithmetic, off-by-one errors can occur.
(https://tc39.es/proposal-temporal/#sec-temporal-parsetemporaldurationstring)

Contrary to the spec, test262 tests had been expecting IEEE-754-aligned results.
Tests were corrected in tc39/test262#3544; this patch corrects our implementation.

* JSTests/test262/expectations.yaml:
Mark 16 test cases as passing.

* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::handleFraction):

Canonical link: https://commits.webkit.org/251809@main
